pkg/catalog/catalog.go

```diff
@@ -1015,58 +1024,151 @@ func (c *Catalog) ListCommits(ctx context.Context, repositoryID string, branch s
 		}
 	}
 
-	// collect commits
-	var commits []*CommitLog
+	paths := params.PathList
+	if len(paths) == 0 {
+		return listCommitsWithoutPaths(it, params)
+	}
 
+	return c.listCommitsWithPaths(ctx, repository, it, params)
+}
+
+func (c *Catalog) listCommitsWithPaths(ctx context.Context, repository *graveler.RepositoryRecord, it graveler.CommitIterator, params LogParams) ([]*CommitLog, bool, error) {
 	// commit/key to value cache - helps when fetching the same commit/key while processing parent commits
 	const commitLogCacheSize = 1024 * 5
-	commitCache, err := simplelru.NewLRU(commitLogCacheSize, nil)
+	commitCache, err := lru.New(commitLogCacheSize)
 	if err != nil {
 		return nil, false, err
 	}
 
-	for it.Next() {
-		v := it.Value()
+	const numReadResults = 3
+	done := atomic.NewBool(false)
+	out := make(chan compareJobResult, numReadResults)
 
-		if len(params.PathList) > 0 {
-			// verify that commit includes a change with object/prefix from PathList
-			if len(v.Parents) != NumberOfParentsOfNonMergeCommit {
-				// skip merge commits
-				continue
-			}
-			pathInCommit, err := c.checkPathListInCommit(ctx, repository, v, params.PathList, commitCache)
-			if err != nil {
-				return nil, false, err
-			}
-			if !pathInCommit {
-				// no object or prefix found skip commit
-				continue
+	createdTasks := atomic.NewInt64(-1)
+
+	// Shared workpool for the workers. 2 designated to create the work and receive the result
+	workerGroup, ctx := c.workpool.GroupContext(ctx)
+	var mgmtGroup multierror.Group
+	paths := params.PathList
+	mgmtGroup.Go(func() error {
+		num := 0
+		for ; it.Next(); num++ {
+			numm := num
+			vall := it.Value()
+			if done.Load() {
+				break
 			}
-		}
+			workerGroup.Submit(func() error {
+				if done.Load() {
+					out <- compareJobResult{num: numm}
```

Comment:
Why do we need that last result? is `done.Load()` mean you are already draining?